src/tiled/command.cpp

```diff
@@ -50,6 +51,46 @@ QString Command::finalCommand() const
     return finalCommand;
 }
 
+QString Command::finalWorkingDir() const
+{
+    QString finalWorkingDir = workingDir;
+
+    MapDocument *mapDoc = DocumentManager::instance()->currentDocument();
+    if (mapDoc) {
+        const QString mapPath = mapDoc->path();
+        finalWorkingDir.replace(QLatin1String("%executablepath"),
+                                QLatin1String("\"%1\"")).arg(executablePath());
+        finalWorkingDir.replace(QLatin1String("%mappath"),
+                                QLatin1String("\"%1\"")).arg(mapPath);
+
+    }
+    return finalWorkingDir;
+}
+
+QString Command::executableFileName() const
+{
+    QString finalCmd = finalCommand();
+    QString executableFilename;
+    QStringList stringsToCheck;
+    stringsToCheck << QLatin1String("'") << QLatin1String("\"") << QLatin1String("\\\"");
+    foreach (const QString &s, stringsToCheck) {
+        if (finalCmd.contains(s)) {
+            executableFilename = finalCmd.mid(1, finalCmd.lastIndexOf(s));
+            return executableFilename;
+        }
+    }
+    executableFilename = finalCmd.mid(0, finalCmd.indexOf(QLatin1String(" ")));
+    return executableFilename;
+}
+
+QString Command::executablePath() const
+{
+    QString executableFilename(executableFileName());
+    if (executableFilename.contains(QDir::separator()))
+        return QFileInfo(executableFilename).path();
```

Comment:
I don't think this check is necessary. I tried the case without separator and it seemed to return ".", which just means current directory and would be a correct return value in this case. So you can write this function like:
```c++
QString Command::executablePath() const
{
    return QFileInfo(executableFileName()).path();
}
```